controlplane/internal/controller/agentcontrolplane_controller.go

```diff
@@ -199,6 +216,70 @@ func (r *AgentControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.Re
 	return ctrl.Result{}, rerr
 }
 
+// Ensures dependencies are deleted before allowing the AgentControlPlane to be deleted
+// Deletes the ClusterDeployment (which deletes the AgentClusterInstall)
+// Machines, InfraMachines, and AgentBootstrapConfigs get auto-deleted when the ACP has a deletion timestamp - this deprovisions the BMH automatically
+// TODO: should we handle watching until all machines & agentbootstrapconfigs are deleted too?
+func (r *AgentControlPlaneReconciler) handleDeletion(ctx context.Context, acp *controlplanev1beta1.AgentControlPlane) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	if controllerutil.ContainsFinalizer(acp, ACP_FINALIZER) {
+		log.Info("ACP contains finalizer, handling deletion")
+		// Delete cluster deployment
+		if acp.Spec.AgentConfigSpec.ClusterDeploymentRef != nil {
+			// Retrieve clusterdeployment and delete
+			cd := &hivev1.ClusterDeployment{}
+			if err := r.Client.Get(ctx, types.NamespacedName{Namespace: acp.Spec.AgentConfigSpec.ClusterDeploymentRef.Namespace, Name: acp.Spec.AgentConfigSpec.ClusterDeploymentRef.Name}, cd); err != nil {
+				if !apierrors.IsNotFound(err) {
+					log.Error(err, "failed getting cluster deployment for ACP")
+					return ctrl.Result{}, err
+				}
+				// Continue if not found
+			} else if err := r.Client.Delete(ctx, cd); err != nil {
+				log.Error(err, "failed deleting cluster deployment for ACP")
+				return ctrl.Result{}, err
+			}
```

Comment:
What's the downside of running delete directly (without running GET previously)? Can we differentiate "not found" on deletion from other errors?

Reply:
Good point! Modified to just delete